src/iarc7_vision/roomba_blob_vision_node.py

```diff
@@ -0,0 +1,351 @@
+
+class CameraProcessor(ImageRoombaFinder):
+
+    def __init__(self, debugging_on):
+        rospy.init_node(NODE_NAME, anonymous=True)
+
+        self.debug = Debugger(debugging_on, debugging_on)
+
+        self.roombas = []
+
+        self.bridge = CvBridge()
+        self.camera = image_geometry.PinholeCameraModel()
+
+        self.tf_buffer = tf2_ros.Buffer()
+        tf2_ros.TransformListener(self.tf_buffer)
+        # Make sure at least one transform exists before starting
+        self.tf_buffer.lookup_transform('map', 'bottom_camera_optical',
+                                        rospy.Time(0), rospy.Duration(3.0))
+
+        rospy.Subscriber("/bottom_image_raw/image", Image, self.callback)
+        rospy.Subscriber("/bottom_image_raw/camera_info", CameraInfo, self.camera.fromCameraInfo)
+        self.publisher = rospy.Publisher("/roombas", OdometryArray, queue_size=10)
+
+        # Starup the loop
+        try:
+            rospy.spin()
+        except KeyboardInterrupt:
+            rospy.loginfo("Shutting down")
+        
+    def callback(self, data):
+        """
+        Receive the callback from the camera, process the image, and publish
+        the roomba locations found
+
+        :param data: Image
+        :type data: sensor_msgs.msg.Image
+        :return: None
+        """
+        try:
+            cv_image = self.bridge.imgmsg_to_cv2(data, "bgr8")
+            # cv_image = cv2.resize(cv_image,None,fx=0.5,fy=0.5, 
+            #                       interpolation=cv2.INTER_CUBIC)
+            trans = self.tf_buffer.lookup_transform('map',
+                                'bottom_camera_optical', rospy.Time(0))
+            self.filter_frame(cv_image, trans)
+        except (tf2_ros.LookupException, tf2_ros.ConnectivityException, \
+                tf2_ros.ExtrapolationException, CvBridgeError) as e:
+            rospy.logerr(e)
+
+
+    def filter_frame(self, frame, trans):
+        """
+        Applies a four-step algorithm to find all the roombas in a given BGR image.
+    
+        :param frame: raw BGR image in which to find roombas
+        :type frame: numpy.ndarray of shape 3 x WIDTH x HEIGHT with dtype uint8
+        :return: None
+    
+        .. note::
+            Does not return data, only debugs the result
+    
+        .. note::
+            The image origin is in the upper-left hand corner
+        """
+        points = []
+    
+        for img_coords in self.bound_roombas(frame):
+            cam_ray = Vector3Stamped()
+            cam_ray.vector.x, cam_ray.vector.y, cam_ray.vector.z = self.camera.projectPixelTo3dRay(img_coords)
+            
+            # Convert that camera ray to world space (map frame)
+            map_ray = tf2_geometry_msgs.do_transform_vector3(cam_ray, copy.deepcopy(trans)) # reassigning
+            
+            # Scale the direction to hit the ground (plane z=0)
+            direction_scale = (trans.transform.translation.z - ROOMBA_HEIGHT) / map_ray.vector.z
```

Comment:
I would typically do this by creating a `Point` at the origin of the `bottom_camera_optical` frame, calling `do_transform_point` to put it in the `map` frame, and then looking at the `z` coordinate of the result, just because I think it's more well-defined.  Same thing right below here.  That way, it's immediately obvious to the reader how the signs of all the numbers should work, and which frame all of the geometric objects are in.